source/zipkin/zipkin_core_types.h

```diff
@@ -0,0 +1,284 @@
+#pragma once
+
+#include <string>
+#include <vector>
+#include <memory>
+
+#include "util.h"
+#include "tracer_interface.h"
+
+namespace Zipkin {
+
+class ZipkinBase {
+public:
+  virtual ~ZipkinBase() {}
+
+  virtual const std::string& toJson() = 0;
+
+protected:
+  std::string json_string_;
+};
+
+class Endpoint : public ZipkinBase {
+public:
+  Endpoint(const Endpoint&);
+
+  Endpoint& operator=(const Endpoint&);
+
+  Endpoint() : ipv4_(), port_(0), service_name_(), isset_ipv6_(false) {}
+
+  Endpoint(std::string& ipv4, uint16_t port, const std::string& service_name)
+      : ipv4_(ipv4), port_(port), service_name_(service_name), isset_ipv6_(false) {}
+
+  const std::string& ipv4() const { return ipv4_; }
+
+  void setIpv4(std::string& ipv4) { ipv4_ = ipv4; }
+
+  const std::string& ipv6() const { return ipv6_; }
+
+  void setIpv6(const std::string& ipv6) {
+    ipv6_ = ipv6;
+    isset_ipv6_ = true;
+  }
+
+  bool isSetIpv6() const { return isset_ipv6_; }
+
+  uint16_t port() const { return port_; }
+
+  void setPort(uint16_t port) { port_ = port; }
+
+  const std::string& serviceName() const { return service_name_; }
+
+  void setServiceName(const std::string& service_name) { service_name_ = service_name; }
+
+  const std::string& toJson() override;
+
+private:
+  std::string ipv4_;
+  uint16_t port_;
+  std::string service_name_;
+  std::string ipv6_;
+
+  bool isset_ipv6_;
+};
+
+class Annotation : public ZipkinBase {
+public:
+  Annotation(const Annotation&);
+
+  Annotation& operator=(const Annotation&);
+
+  Annotation() : timestamp_(0), value_(), isset_host_(false) {}
+
+  Annotation(uint64_t timestamp, const std::string value, Endpoint& host)
+      : timestamp_(timestamp), value_(value), host_(host), isset_host_(true) {}
+
+  const Endpoint& host() const { return host_; }
+
+  void setHost(const Endpoint& host) {
+    host_ = host;
+    isset_host_ = true;
+  }
+
+  void setHost(const Endpoint&& host) {
+    host_ = host;
+    isset_host_ = true;
+  }
+
+  uint64_t timestamp() const { return timestamp_; }
+
+  void setTimestamp(uint64_t timestamp) { timestamp_ = timestamp; }
+
+  const std::string& value() const { return value_; }
+
+  void setValue(const std::string& value) { value_ = value; }
+
+  bool isSetHost() const { return isset_host_; }
+
+  const std::string& toJson() override;
+
+private:
+  uint64_t timestamp_;
+  std::string value_;
+  Endpoint host_;
+
+  bool isset_host_;
+};
+
+enum AnnotationType { BOOL = 0, BYTES = 1, I16 = 2, I32 = 3, I64 = 4, DOUBLE = 5, STRING = 6 };
```

Comment:
you can simplify this by only supporting BOOL and STRING as others are not used in practice in zipkin.

BOOL is a special case for identifying the peer address. for example, "ca", true, host=client address, not the local address.

https://github.com/openzipkin/zipkin-api/blob/master/thrift/zipkinCore.thrift#L226